Support nested temporary listeners

0x0badc0de opened the pull request:
If there are multiple places in a code which use temporary global listeners, the first block w/ listeners which leaves cleans all the other listeners from other blocks as well. Example:

```
def foo
  Wisper.subscribe(MyListener.new) do
    bar
  end
end

def bar
  Wisper.subscribe(MyAnotherListener.new) do
    baz
  end
  broadcast(:something) # won't be catched because listeners in `foo` are cleared away
end
```

krisleech commented:
Nice catch, the PR looks good, thanks. I was going to question the use of `Set.new` but looking at the tests it appears to be used to prevent a listener instance which is temporarily subscribed in both blocks getting events twice - would that be correct?

0x0badc0de commented:
Replaced Set w/ Array (it was used just because `registrations` is a set but `merge` and `subtract` accept any enumerable).
Also updated spec description. It is supposed to check that if we subscribe the same listener twice it won't be unsubscribed when nested block exits (so will receive `:failure`). Hope it is more clear now.

bemurphy commented:
I was looking at the temporary listeners code and didn't spot any fixes to this, am I correct in thinking this is still outstanding?

I believe I just hit this tonight, was real tough to figure out. Fortunately I eventually spotted this issue and it tipped me off.

gondalez commented:
hey @krisleech I have bumped into this issue on a project too.. is it possible to merge this or is it being held out deliberately? 🙏 Thanks for the great lib!

krisleech commented:
@0x0badc0de Are you okay for me to merge https://github.com/krisleech/wisper/pull/175 instead of this one? It looks like you might have deleted the branch.

0x0badc0de commented:
@krisleech sure, go for it
